nats/src/main.rs

```diff
@@ -310,3 +398,55 @@ impl Messaging for NatsMessagingProvider {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use wasmbus_rpc::provider::ProviderHandler;
+
+    /// Ensure that unlink triggers subscription removal
+    /// https://github.com/wasmCloud/capability-providers/issues/196
+    ///
+    /// NOTE: this is tested here for easy access to put_link/del_link without
+    /// the fuss of loading/managing individual actors in the lattice
+    #[tokio::test]
+    async fn test_unlink_unsub() {
```

Comment:
This test doesn't actually test that the subs were deleted unfortunately. To check that the sub was deleted, we'd have to talk to nats. The stuff was always getting removed from the hashmap, but we weren't aborting the subscription threads

Reply:
I think it does -- I restructured `delete_link` to force going through the handles and dropping, so the only way you could get delete_link to finish is by clearing out the subs array. which I explicitly check in the test. 

The only way we could get through unsubbing without going through that code path is if something *else* removed the sub before hand (and somehow did so without causing the `drop` of the handle).

This test does not *double check* with NATS, but lexically I think we can be sure that that codepath is run through, given that no other codepath that calls `aw.remove`. There exists no other way for `aw.remove` to be called without going through the codepath that calls `abort()` and the explicit `drop()` (which is superfluous as you mention).

Does that logic make sense?

Reply:
The error as it was occurring before is that nothing was terminating the spawned process. We didn't store join handles and so couldn't abort on drop. So the client was removed from the hashmap, but that didn't actually terminate the spawned tasks. This test does check that we actually remove things when a link is deleted, so that is fine I think, but it doesn't actually do any checking that the subs actually stopped (which we can only do through checking NATS). I think it is good enough for now

Reply:
yeah, going through and checking NATS is the :medal_sports: ...